File: src/temporal_xray/tools/get_workflow_history.py
from __future__ import annotations

from ..temporal_client import get_temporal_client
from ..transformers.history_summarizer import summarize_history
from ..types import GetWorkflowHistoryInput, WorkflowHistory


async def get_workflow_history(input: GetWorkflowHistoryInput) -> WorkflowHistory:
    client = await get_temporal_client(input.namespace)

    try:
        handle = client.get_workflow_handle(input.workflow_id, run_id=input.run_id)

        history = await handle.fetch_history()
        events = list(history.events) if history.events else []

        if not events:
            raise RuntimeError(
                f"No workflow found with ID '{input.workflow_id}'. "
                "The workflow may have been archived or the ID may be incorrect."
            )

        description = await handle.describe()
        run_id = input.run_id or description.run_id

        return summarize_history(
            input.workflow_id,
            run_id,
            events,
            detail_level=input.detail_level or "summary",
            event_types=input.event_types,
        )
    except RuntimeError:
        raise
    except Exception as e:
        message = str(e)
        if "not found" in message.lower() or "NotFound" in message:
            raise RuntimeError(
                f"No workflow found with ID '{input.workflow_id}'. "
                "The workflow may have been archived or the ID may be incorrect."
            ) from e
        if "permission" in message.lower() or "Unauthorized" in message:
            raise RuntimeError(
                "Permission denied. The configured credentials don't have read access to this namespace."
            ) from e
        raise RuntimeError(f"Failed to get workflow history: {message}") from e

File: src/tools/get-workflow-history.ts
import type { temporal } from "@temporalio/proto";
import { getTemporalClient } from "../temporal-client.js";
import { summarizeHistory } from "../transformers/history-summarizer.js";
import type { GetWorkflowHistoryInput, WorkflowHistory } from "../types.js";

export async function getWorkflowHistory(
  input: GetWorkflowHistoryInput
): Promise<WorkflowHistory> {
  const client = await getTemporalClient();

  try {
    const handle = client.workflow.getHandle(
      input.workflow_id,
      input.run_id
    );

    // Fetch the full event history
    const history = await handle.fetchHistory();
    const events: temporal.api.history.v1.IHistoryEvent[] = history.events || [];

    if (events.length === 0) {
      throw new Error(
        `No workflow found with ID '${input.workflow_id}'. The workflow may have been archived or the ID may be incorrect.`
      );
    }

    // Get the run ID from the describe response
    const description = await handle.describe();
    const runId = input.run_id || description.runId;

    return summarizeHistory(input.workflow_id, runId, events, {
      detailLevel: input.detail_level || "summary",
      eventTypes: input.event_types,
    });
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);

    if (message.includes("not found") || message.includes("NotFound")) {
      throw new Error(
        `No workflow found with ID '${input.workflow_id}'. The workflow may have been archived or the ID may be incorrect.`
      );
    }

    if (message.includes("permission") || message.includes("Unauthorized")) {
      throw new Error(
        `Permission denied. The configured credentials don't have read access to this namespace.`
      );
    }

    throw new Error(`Failed to get workflow history: ${message}`);
  }
}
